rustls-platform-verifier/src/verification/windows.rs

```diff
@@ -214,6 +212,90 @@ impl Drop for Certificate {
+        for root in roots {
+            exclusive_store.add_cert(root)?;
+        }
+
+        let mut config = CERT_CHAIN_ENGINE_CONFIG::zeroed_with_size();
+        config.hExclusiveRoot = exclusive_store.inner.as_ptr();
+
+        let mut engine = 0;
+        // SAFETY: `engine` is valid to be written to and the config is valid to be read.
+        let res = unsafe { CertCreateCertificateChainEngine(&config, &mut engine) };
+
+        #[allow(clippy::as_conversions)]
+        let engine = call_with_last_error(|| match NonNull::new(engine as *mut c_void) {
+            Some(c) if res == TRUE => Some(c),
+            _ => None,
+        })?;
+        Ok(Self { inner: engine })
+    }
+
+    #[cfg(any(test, feature = "ffi-testing", feature = "dbg"))]
+    fn new_with_fake_root(root: &[u8]) -> Result<Self, TlsError> {
+        use windows_sys::Win32::Security::Cryptography::{
+            CERT_CHAIN_CACHE_ONLY_URL_RETRIEVAL, CERT_CHAIN_ENABLE_CACHE_AUTO_UPDATE,
+        };
+
+        let mut root_store = CertificateStore::new()?;
+        root_store.add_cert(root)?;
+
+        let mut config = CERT_CHAIN_ENGINE_CONFIG::zeroed_with_size();
+        // We use these flags for the following reasons:
+        //
+        // - CERT_CHAIN_CACHE_ONLY_URL_RETRIEVAL is used in an attempt to stop Windows from using the internet to
+        // fetch anything during the tests, regardless of what test data is used.
+        //
+        // - CERT_CHAIN_ENABLE_CACHE_AUTO_UPDATE is used as a minor performance optimization to allow Windows to reuse
+        // data inside of a test and avoid any extra parsing, etc, it might need to do pulling directly from the store each time.
+        //
+        // Ref: https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_chain_engine_config
+        config.dwFlags = CERT_CHAIN_CACHE_ONLY_URL_RETRIEVAL | CERT_CHAIN_ENABLE_CACHE_AUTO_UPDATE;
+        config.hExclusiveRoot = root_store.inner.as_ptr();
+
+        let mut engine = 0;
+        // SAFETY: `engine` is valid to be written to and the config is valid to be read.
+        let res = unsafe { CertCreateCertificateChainEngine(&config, &mut engine) };
+
+        #[allow(clippy::as_conversions)]
+        let engine = call_with_last_error(|| match NonNull::new(engine as *mut c_void) {
+            Some(c) if res == TRUE => Some(c),
+            _ => None,
+        })?;
+
+        Ok(Self { inner: engine })
+    }
+
+    #[allow(clippy::as_conversions)]
+    fn as_ptr(&self) -> isize {
+        self.inner.as_ptr() as isize
+    }
+}
+
+impl Drop for CertEngine {
+    fn drop(&mut self) {
+        // SAFETY: The engine pointer is guaranteed to be non-null.
+        unsafe { CertFreeCertificateChainEngine(self.as_ptr()) };
+    }
+}
+
+// SAFETY: We know no other threads is mutating the `CertEngine`, because it would require `unsafe`.
+// Across the FFI, `CertGetCertificateChain` don't mutate it either.
```

Comment:
No action needed: I'm not 100% certain of this assertion FWIW, but I think there's enough supporting evidence to go with it. Windows handles are generally thread safe, the `CertFreeCertificateChainEngine` function doesn't say it invalidates the handle, and an engine is really just a set of parameters Windows needs to read.